### ISSUE TITLE
type context causes failed call to `dict.__init__` with TypedDict argument

### ISSUE THREAD

**Issue by carljm** [astral-sh/ty#3340]
```py
from typing import TypedDict

class TD(TypedDict):
    x: int
    y: str

def takes_kwargs(**kwargs: object): ...
def takes_dict(d: dict[str, object]): ...
    
def _(data: TD):
    reveal_type(dict(data))  # we reveal `dict[Unknown, Unknown]`, other type checkers reveal `dict[str, object]`
    takes_dict(dict(data))  # we error with "no matching overload", other type checkers are fine with this call
    takes_kwargs(**dict(data))  # everyone is fine with this call
```

**Comment by carljm** [comment]
This showed up in hydpy in the ecosystem report on https://github.com/astral-sh/ruff/pull/24653, but it's not related to that feature, it's just that the TypedDict type only started to exist in that location once we supported that feature.

**Comment by charliermarsh** [comment]
Reminds me of https://github.com/astral-sh/ruff/pull/24709 (but not the same).
